mesonbuild/compilers/detect.py

```diff
@@ -0,0 +1,1220 @@
+defaults['cuda'] = ['nvcc']
+defaults['rust'] = ['rustc']
+defaults['swift'] = ['swiftc']
+defaults['vala'] = ['valac']
+defaults['cython'] = ['cython']
+defaults['static_linker'] = ['ar', 'gar']
+defaults['strip'] = ['strip']
+defaults['vs_static_linker'] = ['lib']
+defaults['clang_cl_static_linker'] = ['llvm-lib']
+defaults['cuda_static_linker'] = ['nvlink']
+defaults['gcc_static_linker'] = ['gcc-ar']
+defaults['clang_static_linker'] = ['llvm-ar']
+
+
+def compiler_from_language(env: 'Environment', lang: str, for_machine: MachineChoice) -> T.Optional[Compiler]:
+    lang_map: T.Dict[str, T.Callable[['Environment', MachineChoice], Compiler]] = {
+        'c': detect_c_compiler,
+        'cpp': detect_cpp_compiler,
+        'objc': detect_objc_compiler,
+        'cuda': detect_cuda_compiler,
+        'objcpp': detect_objcpp_compiler,
+        'java': detect_java_compiler,
+        'cs': detect_cs_compiler,
+        'vala': detect_vala_compiler,
+        'd': detect_d_compiler,
+        'rust': detect_rust_compiler,
+        'fortran': detect_fortran_compiler,
+        'swift': detect_swift_compiler,
+        'cython': detect_cython_compiler,
+    }
+    return lang_map[lang](env, for_machine) if lang in lang_map else None
+
+def detect_compiler_for(env: 'Environment', lang: str, for_machine: MachineChoice)-> T.Optional[Compiler]:
+    comp = compiler_from_language(env, lang, for_machine)
+    if comp is not None:
+        assert comp.for_machine == for_machine
+        env.coredata.process_new_compiler(lang, comp, env)
+    return comp
+
+
+# Helpers
+# =======
+
+def _get_compilers(env: 'Environment', lang: str, for_machine: MachineChoice) -> T.Tuple[T.List[T.List[str]], T.List[str], T.Optional['ExternalProgram']]:
+    '''
+    The list of compilers is detected in the exact same way for
+    C, C++, ObjC, ObjC++, Fortran, CS so consolidate it here.
+    '''
+    value = env.lookup_binary_entry(for_machine, lang)
+    if value is not None:
+        comp, ccache = BinaryTable.parse_entry(value)
+        # Return value has to be a list of compiler 'choices'
+        compilers = [comp]
+    else:
+        if not env.machines.matches_build_machine(for_machine):
+            raise EnvironmentException(f'{lang!r} compiler binary not defined in cross or native file')
+        compilers = [[x] for x in defaults[lang]]
+        ccache = BinaryTable.detect_ccache()
+
+    if env.machines.matches_build_machine(for_machine):
+        exe_wrap: T.Optional[ExternalProgram] = None
+    else:
+        exe_wrap = env.get_exe_wrapper()
+
+    return compilers, ccache, exe_wrap
+
+def _handle_exceptions(
+        exceptions: T.Mapping[str, T.Union[Exception, str]],
+        binaries: T.List[T.List[str]],
+        bintype: str = 'compiler'
+    ) -> T.NoReturn:
+    errmsg = f'Unknown {bintype}(s): {binaries}'
+    if exceptions:
+        errmsg += '\nThe following exception(s) were encountered:'
+        for (c, e) in exceptions.items():
```

Comment:
style nit, you don't need the parens here. You only need them when you're doing complex unpacking, such as a nested tuple

Reply:
Fixed